### PR TITLE
feat: Migrate S3 backend to AWS SDK for Go v2

### LINKED ISSUE #30443: backend/s3: Upgrade to `aws-sdk-go-base/v2`
The AWS provider will be upgrading to [`aws-sdk-go-base/v2`](https://github.com/hashicorp/aws-sdk-go-base) as part of v4.0. Upgrading to `aws-sdk-go-base/v2` will provide enhancements to the AWS authentication process and add consistency to the configuration of both the backend and the provider.

Note: Because `aws-sdk-go-base/v2` changes the authentication flow to be more in-line with the default flow for the AWS CLI and SDKs, this will be a multi-stage process over several releases of Terraform to prevent breaking changes. It will also require changes to `aws-sdk-go-base/v2` itself.

### LINKED ISSUE #32448: error configuring S3 Backend: no valid credential sources for S3 Backend found.
### Terraform Version

```shell
1.3.6
```

### AWS CLI Version

```shell
aws-cli/2.9.12 Python/3.11.1 Darwin/22.2.0 source/arm64 prompt/off
```

### Terraform Configuration Files

```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.30.0"
    }
  }
  backend "s3" {
    bucket         = "terraform-state-int"
    key            = "key/"
    region         = aws-region
    profile = "int"
  }
  required_version = ">= 1.1.0"
}
```

### Debug Output

No relevant debug output except the few lines below (Actual Behaviour)

### Expected Behavior

Should have applied initial config and download plugins

### Actual Behavior

Error: error configuring S3 Backend: no valid credential sources for S3 Backend found.
│ 
│ Please see https://www.terraform.io/docs/language/settings/backends/s3.html
│ for more information about providing credentials.
│ 
│ Error: SSOProviderInvalidToken: the SSO session has expired or is invalid
│ caused by: open /Users/rsharma/.aws/sso/cache/9ceab8b92e009ecc15fdc2be3fe58963c97c5870.json: no such file or directory
│ 

### Steps to Reproduce

1. terraform init

### Additional Context

I have set up and created aws sso profile. I can access aws s3 buckets using that profile ` aws s3 ls --profile PROFILE_NAME`. I can also see json files in .aws/sso/cache and .aws/cli/cache. However, with terraform init, it couldn't file the the particular session

### References

_No response_

### LINKED ISSUE #32465: AWS S3 state backend fails with AWS SSO new profile format
### Terraform Version

```shell
Terraform v1.3.7
on darwin_arm64
+ provider registry.terraform.io/datadog/datadog v3.19.1
+ provider registry.terraform.io/hashicorp/aws v3.75.2
+ provider registry.terraform.io/hashicorp/null v3.2.1
```

### Terraform Configuration Files

```terraform
terraform {
  required_version = "~>1.0"

  required_providers {
    aws = {
      version = "~>3.0"
      source  = "hashicorp/aws"
    }
    null = {
      version = "~>3.0"
      source  = "hashicorp/null"
    }
    datadog = {
      source  = "DataDog/datadog"
      version = "~> 3.18"
    }
  }

  backend "s3" {
    region         = "eu-central-1"
    bucket         = "xxxx-bucket-xxxx"
    key            = "state-key"
    dynamodb_table = "lock-table"
    encrypt        = "true"
  }
}
```

### Debug Output

```
...
2023-01-05T17:27:57.944+0100 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2023-01-05T17:27:57.945+0100 [INFO]  CLI command args: []string{"apply"}
2023-01-05T17:27:57.954+0100 [TRACE] Meta.Backend: built configuration for "s3" backend with hash value 3591387704
2023-01-05T17:27:57.955+0100 [TRACE] Preserving existing state lineage "8e30792b-97ac-5674-d17a-955b1985e845"
2023-01-05T17:27:57.955+0100 [TRACE] Preserving existing state lineage "8e30792b-97ac-5674-d17a-955b1985e845"
2023-01-05T17:27:57.956+0100 [TRACE] Meta.Backend: working directory was previously initialized for "s3" backend
2023-01-05T17:27:57.956+0100 [TRACE] Meta.Backend: using already-initialized, unchanged "s3" backend configuration
2023-01-05T17:27:57.959+0100 [INFO]  Attempting to use session-derived credentials
╷
│ Error: error configuring S3 Backend: Error creating AWS session: profile "xxx" is configured to use SSO but is missing required configuration: sso_region, sso_start_url
│ 
│ 
╵
```

### Expected Behavior

Terraform should work with the new AWS config format

### Actual Behavior

Terraform throws error with the new AWS config format

### Steps to Reproduce

1. Configure S3 remote state backend (including creating the bucket, making sure your AWS permissions are sufficient, etc.)
2. Configure your AWS client to use AWS IAM Identity Center (SSO) new format. Example ~/.aws/config entry:
```
[profile xxx]
sso_session = my-sso
sso_account_id = 111122223333
sso_role_name = SampleRole

[sso-session my-sso]
sso_region = us-east-1
sso_start_url = https://my-sso-portal.awsapps.com/start
sso_registration_scopes = sso:account:access
```
> reference: see [here](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sso.html)
3. Login with `AWS_PROFILE=xxx aws sso login`
4. confirm that your bucket is accessible (e.g. `AWS_PROFILE=xxx AWS_REGION=yyy aws s3 ls <your-bucket-name>`)
5. Run e.g. `terraform init`, and receive the error message above.

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by siketyan**
Fixes #30443
Fixes #32448 
Fixes #32465

Since this will be my first-time contribution for Terraform, please let me know if I am missing something to do, or there are some reason(s) that are blocking migration of aws-sdk-go.

## Target Release

1.6.x?

### ENHANCEMENTS

- Internally migrated S3 remote state backend to AWS SDK for Go v2.

### CODE REVIEW

**Comment by siketyan** [comment]
I've tested on my local environment and it worked.
Authentication using AWS Identity Centre (formerly AWS SSO) also works!

File: internal/backend/remote-state/s3/backend_state.go [hunk 1]
```diff
@@ -29,25 +30,29 @@ func (b *Backend) Workspaces() ([]string, error) {
...
-	err := b.s3Client.ListObjectsPages(params, func(page *s3.ListObjectsOutput, lastPage bool) bool {
+	paginator := s3.NewListObjectsV2Paginator(b.s3Client, params)
+	for paginator.HasMorePages() {
+		page, err := paginator.NextPage(context.TODO())
+		if err != nil {
+			var noSuchBucket *types.NoSuchBucket
+			if errors.As(err, &noSuchBucket) {
+				return nil, fmt.Errorf(errS3NoSuchBucket, err)
+			}
+		}
```
Comment (siketyan) [inline internal/backend/remote-state/s3/backend_state.go:48]: What should we do here if an unexpected error is occurred? Previous implementation does not have any error handling for this too, and that causes a panic.

File: internal/backend/remote-state/s3/backend_test.go [hunk 2]
```diff
@@ -55,8 +58,9 @@ func TestBackendConfig(t *testing.T) {
 	}
 
 	b := backend.TestBackendConfig(t, New(), backend.TestWrapConfig(config)).(*Backend)
+	options := reflect.ValueOf(*b.s3Client).Elem().FieldByName("options").Interface().(s3.Options)
```
Comment (siketyan) [inline internal/backend/remote-state/s3/backend_test.go:61]: Are there any options to so this in much better way?

**Comment by gdavison** [comment]
Thanks for submitting this PR, @siketyan. Because `aws-sdk-go-base/v2` changes the authentication flow to match the default behaviour of the AWS CLI and SDKs, simply replacing it would be a breaking change, so we need to roll out the behaviour change over several releases of Terraform.

We have an active project to update this, and we will try to use your changes where we can.

**Comment by siketyan** [comment]
aws-sdk-go v1 has backport to support AWS Identity Center now, and the SDK used in Terraform S3 backend will be updated in #33522.

**Comment by jar-b** [comment]
Hi @siketyan 👋 - thanks for your effort on this! 

As @gdavison mentioned earlier, we've been working on a broader effort to modernize the S3 backend, which will occur over several minor releases. #33730 covered much of the scope of this PR and will be released with Terraform 1.6, so I believe this can now be closed. Thanks again for your engagement!

**Pull request closed without merging**